Framework/Geometry/inc/MantidGeometry/Objects/BeamlineRayTracer.h

```diff
@@ -0,0 +1,335 @@
+#ifndef MANTID_GEOMETRY_BEAMLINERAYTRACER_H_
+#define MANTID_GEOMETRY_BEAMLINERAYTRACER_H_
+
+#include "MantidGeometry/IComponent.h"
+#include "MantidGeometry/Instrument/ComponentInfo.h"
+#include "MantidGeometry/Objects/Track.h"
+#include "MantidKernel/Matrix.h"
+#include "MantidKernel/Quat.h"
+#include "MantidKernel/V3D.h"
+
+#include <deque>
+#include <iterator>
+#include <list>
+
+/**
+BeamlineRayTracer contains a set of free functions that are responsible for
+tracking rays and accumulating a list of objects that are intersected along the
+way. These methods have been adapted from the original InstrumentRayTracer.
+
+@author Bhuvan Bezawada, STFC
+@date 17/08/2018
+
+Copyright &copy; 2018 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+National Laboratory & European Spallation Source
+
+This file is part of Mantid.
+
+Mantid is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3 of the License, or
+(at your option) any later version.
+
+Mantid is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+File change history is stored at: <https://github.com/mantidproject/mantid>
+Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+
+namespace Mantid {
+namespace Geometry {
+namespace BeamlineRayTracer {
+
+using Kernel::Matrix;
+using Kernel::Quat;
+using Kernel::V3D;
+using Links = Track::LType;
+using Types = Mantid::Beamline::ComponentType;
+
+namespace IntersectionHelpers {
+
+/**
+ * Tests the intersection of the ray with a rectangular bank of detectors.
+ * Uses the knowledge of the RectangularDetector shape to significantly speed up
+ * tracking.
+ *
+ * @param track :: Track under test. The results are stored here.
+ */
+void checkIntersectionWithRectangularBank(Track &track,
```

Comment:
Why is all the implementation in the header? Separate declaration from definition. Put definition into cpp file.